3.assess_generalizability/scripts/generalizabilty_eval.py

```diff
@@ -0,0 +1,324 @@
+    """
+    # Get the feature columns (excluding 'Metadata' columns and the label column)
+    feature_columns = [
+        col for col in df.columns if not col.startswith("Metadata") and col != label
+    ]
+
+    # Extract feature space (X) as a DataFrame (keep feature names)
+    X = df[feature_columns]
+
+    # Extract class label (y) as a NumPy array
+    y = df.loc[:, [label]].values
+    y = np.ravel(y)  # Flatten y to a 1D array
+
+    # If shuffle is True, shuffle the feature columns independently for training
+    if shuffle:
+        for column in X.T:
+            np.random.shuffle(column)
+
+    return X, y
+
+
+# ## Set paths and variables
+
+# In[ ]:
+
+
+# Path to folder holding model and encoder files
+model_dir = pathlib.Path("../1.train_models/data")
+
+# Path to results directory
+results_dir = pathlib.Path("./results")
+results_dir.mkdir(exist_ok=True)
+
+# Load in the model encoder
+le = load(pathlib.Path(f"{model_dir}/trained_nf1_model_label_encoder.joblib"))
+
+# Load in the model
+model = load(pathlib.Path(f"{model_dir}/trained_nf1_model.joblib"))
+
+# Set the random seed
+rng = np.random.default_rng(0)
+
+
+# ## Load in plate with two cell lines (Plate 6)
+
+# In[4]:
+
+
+# Read in data from plate 6 with two cell lines
+plate6_df = pd.read_parquet(
+    pathlib.Path(
+        "/media/18tbdrive/1.Github_Repositories/nf1_schwann_cell_painting_data/3.processing_features/data/single_cell_profiles/Plate_6_sc_normalized.parquet"
+    )
+)
+
+# Remove rows where Metadata_genotype is "HET"
+plate6_df = plate6_df[plate6_df["Metadata_genotype"] != "HET"].reset_index(drop=True)
+
+# Count rows before dropping NaNs
+initial_count = plate6_df.shape[0]
+
+# Drop rows with NaNs
+plate6_df = plate6_df.dropna()
+
+# Count rows after dropping NaNs
+final_count = plate6_df.shape[0]
+
+# Print the count of dropped rows
+print(f"Dropped rows: {initial_count - final_count}")
+
+# Print shape and head of data
+print(plate6_df.shape)
+plate6_df.head()
+
+
+# ## Generate a shuffled dataset from the loaded in plate
+
+# In[5]:
+
+
+# Shuffle the features randomly, excluding columns that start with "Metadata_"
+shuffled_plate6_df = plate6_df.apply(
+    lambda x: rng.permutation(x) if not x.name.startswith("Metadata_") else x
+)
+
+# Print shape and head of data
+print(shuffled_plate6_df.shape)
+shuffled_plate6_df.head()
+
+
+# ## Apply model to final and shuffled versions of the plate data
+
+# In[6]:
+
+
+# Create list of the metadata columns only
+meta_cols = [col for col in plate6_df.columns if "Metadata" in col]
+
+# Define a dictionary to handle both data types
+data_dict = {"final": plate6_df, "shuffled": shuffled_plate6_df}
+
+# Initialize a list to store processed dataframes
+processed_dfs = []
+
+# Loop through the data dictionary to create probability dataframes
+for data_type, data in data_dict.items():
+    # Drop rows with Metadata_genotype == "HET"
+    data = data[data["Metadata_genotype"] != "HET"]
```

Comment:
My recommendation would be to apply the model to everything, and then at eval and figure time drop HETs

Reply:
Will do!